compiler/src/dmd/expressionsem.d

```diff
@@ -13412,19 +13412,98 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         Type t1 = exp.e1.type.toBasetype();
         Type t2 = exp.e2.type.toBasetype();
 
-        // Indicates whether the comparison of the 2 specified array types
-        // requires an object.__equals() lowering.
-        static bool needsDirectEq(Type t1, Type t2, Scope* sc)
+        static bool unifyArrayTypes(Type t1, Type t2, Scope* sc)
         {
             Type t1n = t1.nextOf().toBasetype();
             Type t2n = t2.nextOf().toBasetype();
+
             if ((t1n.ty.isSomeChar && t2n.ty.isSomeChar) ||
                 (t1n.ty == Tvoid || t2n.ty == Tvoid))
             {
-                return false;
+                return true;
             }
             if (t1n.constOf() != t2n.constOf())
+            {
+                return false;
+            }
+
+            Type t = t1n;
+            while (t.toBasetype().nextOf())
+                t = t.nextOf().toBasetype();
+            if (auto ts = t.isTypeStruct())
+            {
+                // semanticTypeInfo() makes sure hasIdentityEquals has been computed
+                if (global.params.useTypeInfo && Type.dtypeinfo)
+                    semanticTypeInfo(sc, ts);
+
+                return !ts.sym.hasIdentityEquals; // has custom opEquals
+            }
+
+            return true;
+        }
+
+        extern(C++) static final class IsMemcmpableVisitor : Visitor
+        {
+            alias visit = Visitor.visit;
+            public:
+            bool result = false;
+
+            override void visit(Type t)
+            {
+                result = t.ty == Tvoid || (t.isScalar() && !t.isFloating());
```

Comment:
```suggestion
                result = t.ty == Tvoid || t.isIntegral();
```
https://github.com/dlang/dmd/blob/master/compiler/src/dmd/astbase.d#L3578-L3581

Reply:
Indeed, the only difference I see is that raw pointers are also scalars, which means they should be `memcmp()`'d when there's no `opEquals`. `isScalar()` it is then.